src/main/scala/is/hail/asm4s/FunctionBuilder.scala

```diff
@@ -130,8 +131,51 @@ class MethodBuilder(val fb: FunctionBuilder[_], val mname: String, val parameter
   }
 }
 
+trait DependentFunction[F >: Null <: AnyRef] extends FunctionBuilder[F] {
+  var definedFields: ArrayBuilder[Growable[AbstractInsnNode] => Unit] = new ArrayBuilder(16)
+
+  def addField[T : TypeInfo : ClassTag](value: Code[T]): ClassFieldRef[T] = {
+    val cfr = newField[T]
+    val add: (Growable[AbstractInsnNode]) => Unit = { (il: Growable[AbstractInsnNode]) =>
+      il += new TypeInsnNode(CHECKCAST, name)
+      value.emit(il)
+      il += new FieldInsnNode(PUTFIELD, name, cfr.name, typeInfo[T].name)
+    }
+    definedFields += add
+    cfr
+  }
+
+  def newInstance()(implicit fct: ClassTag[F]): Code[F] = {
```

Comment:
Excellent. One last bit: make this a type bound rather than an implicit parameter.

Reply:
(womp, that actually didn't work so I reverted it.)